### PR TITLE
Remove support for legacy Sphinx payloads

### PR DESCRIPTION
**Pull request by t-bast**
We previously supported a 65-bytes fixed-size sphinx payload, which has been deprecated in favor of variable length payloads containing a tlv stream (see https://github.com/lightning/bolts/pull/619).

It looks like the whole network now supports the variable-length format, so we can stop accepting the old one. It is also being removed from the spec (see https://github.com/lightning/bolts/pull/962).

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala:92]: It would probably be more fool proof to make the method return a `Try[Int]` instead of relying on the caller to be careful, but that should probably be done in its own PR.

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala:256]: I'm confused, how come all `FinalPayload` are `TrampolinePacket`?

**Review by pm47 — COMMENTED** [review]
That is some serious cleanup 🧹 💨

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala:256]: We use these traits (`PaymentPacket` and `TrampolinePacket`) as "this could be this packet type", not "it has to be this packet type". I agree that this is a bit confusing (it was already the case before but it was hidden by a more complex trait hierarchy) but it's quite handy! I'll try to think of an alternative to clean that up.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala:92]: I'll give this a try, it may not be that much more to add, in which case it makes sense to do now.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala:92]: I does change quite a lot the whole Sphinx file...I think for the best, but I'd rather do that in a separate PR.

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala:256]: This hierarchy is quite handy, because the way we use the `PacketType` is to restrict what kinds of payloads can be used in each context (trampoline or normal payment). I think it works well, but if you have a better alternative let me know!

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala:256]: Maybe something like that?

```scala
  sealed trait FinalPayload extends PerHopPayload {
    val records: TlvStream[OnionPaymentPayloadTlv]
    val amount = records.get[AmountToForward].get.amount
    ...
    val paymentMetadata = records.get[PaymentMetadata].map(_.data)
  }

  case class PaymentFinalPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends FinalPayload with PaymentPacket

  case class TrampolineFinalPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends FinalPayload with TrampolinePacket
```

Didn't dig much, and that would be for another PR anyway.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala:256]: Agreed, I'll dig into this more in a follow-up PR, along with the improvements in `Sphinx.scala`

**Review by pm47 — DISMISSED** [review]

**Comment by t-bast** [comment]
Thanks for the reviews! Actually we should probably hold off on merging this, when looking at the `crypto.sphinx.onion-payload-format` it looks like a non-negligible portion of the payments we relay are still using the old format...so there may be an implementation out there that supports the new format but still encodes with the legacy one (as eclair did earlier)?
I'll wait for feedback from other implementations before merging, we don't want to lose too much relay fees by deploying this.

**Review by pm47 — APPROVED** [review]

**Pull request merged**
